gateleen-validation/src/main/java/org/swisspush/gateleen/validation/Validator.java

```diff
@@ -123,65 +157,71 @@ public static ValidationResult validateStatic(Buffer dataToBeValidated, String s
             log.warn(message);
             return new ValidationResult(ValidationStatus.VALIDATED_NEGATIV, message);
         }
     }
 
-    private static void performValidation(String dataString, JsonObject data, Logger log, String base, Buffer jsonBuffer, String type, String path, Handler<ValidationResult> callback) {
-        if(SCHEMA_DECLARATION.equals(data.getString("$schema"))) {
+    private static void performValidation(JsonSchema schema, Logger log, String base, Buffer jsonBuffer, String type,
+                                          String path, Handler<ValidationResult> callback) {
+        try {
+            JsonNode jsonNode = new ObjectMapper().readTree(jsonBuffer.getBytes());
+            if (jsonNode == null) {
+                throw new IOException("no vaild JSON object: " + jsonBuffer.toString());
+            }
+            final Set<ValidationMessage> valMsgs = schema.validate(jsonNode);
+            if (valMsgs.isEmpty()) {
+                log.debug("Valid ({})", type);
+                log.debug("Used schema: {}", base);
+                callback.handle(new ValidationResult(ValidationStatus.VALIDATED_POSITIV));
+            } else {
+                JsonArray validationDetails = extractMessagesAsJson(valMsgs, log);
+                String messages = StringUtils.getStringOrEmpty(extractMessages(valMsgs));
+                StringBuilder msgBuilder = new StringBuilder();
+                msgBuilder.append("Invalid JSON for ")
+                        .append(path).append(" (")
+                        .append(type).append("). Messages: ")
+                        .append(messages)
+                        .append(" | Report: ").append(getReportAsString(valMsgs));
+
+                if (log.isDebugEnabled()) {
+                    msgBuilder.append(" | Validated JSON: ").append(jsonBuffer.toString());
+                }
+
+                log.warn(msgBuilder.toString());
+
+                callback.handle(new ValidationResult(ValidationStatus.VALIDATED_NEGATIV, msgBuilder.toString(), validationDetails));
+                log.warn("Used schema: {}", base);
+            }
+        } catch (IOException e) {
+            String message = "Cannot read JSON " + " (" + type + ")";
+            log.warn(message, e.getMessage());
+            callback.handle(new ValidationResult(ValidationStatus.VALIDATED_NEGATIV, message));
+        }
+    }
+
+    private static void performValidation(String dataString, JsonObject data, Logger log, String base, Buffer jsonBuffer,
+                                          String type, String path, Handler<ValidationResult> callback) {
+        if (SCHEMA_DECLARATION.equals(data.getString("$schema"))) {
             JsonSchema schema;
             try {
-                schema = JsonSchemaFactory.getInstance().getSchema(dataString);
+                schema = JsonSchemaFactory.getInstance(SpecVersion.VersionFlag.V4).getSchema(dataString);
             } catch (Exception e) {
```

Comment:
[Pokémon Exception Handling](https://blog.codinghorror.com/new-programming-jargon/#2)

Would it be enough to catch `RuntimeException` or similar?

Reply:
Good point. I change it to `RuntimeException` because looking at the `JsonSchemaFactory` class we could get `IllegalArgumentException` and `JsonSchemaException`

Reply:
There are lots of places in gateleen were we do catch `Exception`. This should be targetet too.